### PR TITLE
Extract search logic from nfts part2

### PR DESCRIPTION
**Pull request by banklesss**
Related to [YOMO-502](https://emurgo.atlassian.net/browse/YOMO-502)

[YOMO-502]: https://emurgo.atlassian.net/browse/YOMO-502?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/Nfts/Nfts.tsx [hunk 1]
```diff
@@ -4,23 +4,31 @@ import {RefreshControl, ScrollView, StyleSheet, Text, View} from 'react-native'
...
 
-  const {search, nfts, isLoading, refetch, isError} = useFilteredNfts({
+  const {isLoading, nfts, refetch, isError} = useNfts(wallet, {
     onSettled: () => {
       if (isManualRefreshing) setIsManualRefreshing(false)
     },
   })
 
+  const {search: nftsSearchTerm} = useSearch()
+  const nftsSearchResult = filterNfts(nftsSearchTerm, nfts).sort((NftA, NftB) => NftA.name.localeCompare(NftB.name))
```
Comment (stackchain) [inline src/Nfts/Nfts.tsx]: save capitals for classes/modules, still I'll be fine with `a,b`.
```suggestion
  const nftsSearchResult = filterNfts(nftsSearchTerm, nfts).sort((nftA, nftB) => nftA.name.localeCompare(nftB.name))
```

Comment (stackchain) [reply src/Nfts/Nfts.tsx]: sorting after searching, implies that only filtered nfts will be sorted, I'd suggest splitting into `sortedNfts` and applying the filter on top of the sorted nfts, or even sorted right before adding the nfts into the data for the list.

File: src/Nfts/Nfts.tsx [hunk 2]
```diff
@@ -37,12 +45,12 @@ export const Nfts = () => {
...
     return (
       <ScreenWrapper>
-        <LoadingScreen nftsCount={nfts.length} />
+        <LoadingScreen nftsCount={nftsSearchResult.length} />
       </ScreenWrapper>
     )
   }
 
-  if (search.length > 0 && nfts.length === 0) {
+  if (nftsSearchTerm.length > 0 && nftsSearchResult.length === 0) {
```
Comment (stackchain) [inline src/Nfts/Nfts.tsx]: const `hasEmptySearchResult` ?

File: src/Nfts/filterNfts.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import {YoroiNft} from '../yoroi-wallets'
+
+export const filterNfts = (nftsSearchTerm: string, nfts: YoroiNft[]): YoroiNft[] => {
```
Comment (stackchain) [inline src/Nfts/filterNfts.ts]: `search: string` or `term: string`, `searchTerm`  would be fine for me, the function name is `filterNfts` so I think nft its implied

File: src/Nfts/filterNfts.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+  const searchTermLowerCase = nftsSearchTerm.toLowerCase()
+  const filteredNfts =
+    searchTermLowerCase.length > 0 && nfts.length > 0
```
Comment (stackchain) [inline src/Nfts/filterNfts.ts]: I think `&& nfts.leng..` is not necessary

**Review by stackchain — COMMENTED** [review]
⬆️

File: src/Nfts/Nfts.tsx (hunk 1 above)
Comment (wolverineks) [reply src/Nfts/Nfts.tsx]: why sort nfts that are not displayed?

File: src/Nfts/Nfts.tsx [hunk 5]
```diff
@@ -4,23 +4,34 @@ import {RefreshControl, ScrollView, StyleSheet, Text, View} from 'react-native'
...
-  const {search, nfts, isLoading, refetch, isError} = useFilteredNfts({
+  const {isLoading, nfts, refetch, isError} = useNfts(wallet, {
     onSettled: () => {
       if (isManualRefreshing) setIsManualRefreshing(false)
     },
   })
 
+  const {search: nftsSearchTerm} = useSearch()
+  const sortedNfts = sortNfts(nfts)
+  const nftsSearchResult = filterNfts(nftsSearchTerm, sortedNfts)
```
Comment (wolverineks) [inline src/Nfts/Nfts.tsx:30]: extract the functions used to sort/filter, but don't wrap the sort/filter. something like:
`const filteredNfts = nfts. filter(match(search))`
`const sortedNfts = nfts.sort(byName)`

File: src/Nfts/Nfts.tsx [hunk 6]
```diff
@@ -4,23 +4,34 @@ import {RefreshControl, ScrollView, StyleSheet, Text, View} from 'react-native'
...
+
+  const handleNftSelect = (index: number) => navigateTo.nftDetails(nftsSearchResult[index].id)
```
Comment (wolverineks) [inline src/Nfts/Nfts.tsx]: don't use index

File: src/Nfts/Nfts.tsx (hunk 5 above)
Comment (wolverineks) [reply src/Nfts/Nfts.tsx:30]: filter then sorry

File: src/Nfts/Nfts.tsx (hunk 1 above)
Comment (stackchain) [reply src/Nfts/Nfts.tsx]: yap, that makes sense, sort for last right before adding to the list component.

File: src/Nfts/Nfts.tsx (hunk 6 above)
Comment (stackchain) [reply src/Nfts/Nfts.tsx]: yap, I think it can receive the `id` directly.

File: src/Nfts/Nfts.tsx [hunk 7]
```diff
@@ -68,7 +79,7 @@ export const Nfts = () => {
             message={strings.noNftsInWallet}
             heading={
               <View>
```
Comment (wolverineks) [inline src/Nfts/Nfts.tsx:81]: why does NoNFTsScreen need an nft count? shouldn't it always be 0?

Comment (wolverineks) [reply src/Nfts/Nfts.tsx:81]: "*Screens" should be directly used in navigators, not inside other components

File: src/Nfts/Nfts.tsx (hunk 6 above)
Comment (banklesss) [reply src/Nfts/Nfts.tsx]: what name would you suggest? it is the index of the nfts array

Comment (wolverineks) [reply src/Nfts/Nfts.tsx]: not the name, i mean literally, dont use indexes

Comment (wolverineks) [reply src/Nfts/Nfts.tsx]: `onSelect(index: number)` -> `onSelect(id: string)`

Comment (wolverineks) [reply src/Nfts/Nfts.tsx]: the selection process is `nft -> index -> nft -> id`
just pass the `id`

**Review by stackchain — APPROVED** [review]
lgtm

File: src/Nfts/Nfts.tsx [hunk 8]
```diff
@@ -4,61 +4,72 @@ import {RefreshControl, ScrollView, StyleSheet, Text, View} from 'react-native'
...
   })
 
+  const {search: nftsSearchTerm} = useSearch()
+  const filteredNfts = filterNfts(nftsSearchTerm, nfts)
+  const sortedNfts = filteredNfts.sort(sortNfts)
+  const nftsSearchResult = filterNfts(nftsSearchTerm, sortedNfts)
+
+  const hasEmptySearchResult = nftsSearchTerm.length > 0 && nftsSearchResult.length === 0
+  const hasNotNfts = nftsSearchTerm.length === 0 && nftsSearchResult.length === 0
+
```
Comment (wolverineks) [inline src/Nfts/Nfts.tsx:34]: why `useCallback`

File: src/Nfts/Nfts.tsx [hunk 9]
```diff
@@ -4,61 +4,72 @@ import {RefreshControl, ScrollView, StyleSheet, Text, View} from 'react-native'
...
     },
   })
 
+  const {search: nftsSearchTerm} = useSearch()
+  const filteredNfts = filterNfts(nftsSearchTerm, nfts)
+  const sortedNfts = filteredNfts.sort(sortNfts)
+  const nftsSearchResult = filterNfts(nftsSearchTerm, sortedNfts)
+
+  const hasEmptySearchResult = nftsSearchTerm.length > 0 && nftsSearchResult.length === 0
+  const hasNotNfts = nftsSearchTerm.length === 0 && nftsSearchResult.length === 0
```
Comment (wolverineks) [inline src/Nfts/Nfts.tsx]: why do you include search term when the wallet doesn't have any nfts?

File: src/Nfts/Nfts.tsx [hunk 10]
```diff
@@ -68,35 +79,40 @@ export const Nfts = () => {
...
     )
   }
 
   return (
-    <ScreenWrapper>
+    <Wrapper>
       <View style={styles.galleryContainer}>
-        {search.length === 0 && (
+        {nftsSearchTerm.length === 0 && (
           <View>
```
Comment (wolverineks) [inline src/Nfts/Nfts.tsx]: `<View>` -> `<>`

File: src/Nfts/Nfts.tsx [hunk 11]
```diff
@@ -140,7 +156,7 @@ function ErrorScreen({onRefresh, isRefreshing}: {onRefresh: () => void; isRefres
   )
 }
 
-function NftCount({count}: {count?: number | string}) {
+const NftCount = ({count}: {count?: number | string}) => {
```
Comment (wolverineks) [inline src/Nfts/Nfts.tsx:159]: inline countText

Comment (wolverineks) [reply src/Nfts/Nfts.tsx:159]: redundant `<View>`

File: src/Nfts/Nfts.tsx (hunk 9 above)
Comment (wolverineks) [reply src/Nfts/Nfts.tsx]: if the wallet doesn't have any nfts, don't show search

File: src/Nfts/Nfts.tsx [hunk 12]
```diff
@@ -4,61 +4,72 @@ import {RefreshControl, ScrollView, StyleSheet, Text, View} from 'react-native'
...
-        <LoadingScreen nftsCount={nfts.length} />
-      </ScreenWrapper>
+      <Wrapper>
+        <LoadingScreen nftsCount={nftsSearchResult.length} />
+      </Wrapper>
     )
   }
 
-  if (search.length > 0 && nfts.length === 0) {
+  if (hasEmptySearchResult) {
```
Comment (wolverineks) [inline src/Nfts/Nfts.tsx:56]: (isError, isLoading) = 1 scope
(hasNotNfts, hasEmptySearchResults) = different scope
don't mix scopes

File: src/Nfts/Nfts.tsx [hunk 13]
```diff
@@ -165,6 +181,8 @@ function LoadingScreen({nftsCount}: {nftsCount: number}) {
   )
 }
 
+const sortNfts = (nftA: YoroiNft, nftB: YoroiNft): number => nftA.name.localeCompare(nftB.name)
```
Comment (wolverineks) [inline src/Nfts/Nfts.tsx]: don't pass in specific objects
pass in primitives
make it generic

File: src/Nfts/Nfts.tsx (hunk 8 above)
Comment (banklesss) [reply src/Nfts/Nfts.tsx:34]: could you specify please?

File: src/Nfts/Nfts.tsx (hunk 12 above)
Comment (banklesss) [reply src/Nfts/Nfts.tsx:56]: What would be the implementation that you suggest? Could you specify?

Thanks!

File: src/Nfts/Nfts.tsx (hunk 9 above)
Comment (banklesss) [reply src/Nfts/Nfts.tsx]: > if the wallet doesn't have any nfts, don't show search

 I think this is something to be considered by product

File: src/Nfts/Nfts.tsx (hunk 8 above)
Comment (banklesss) [reply src/Nfts/Nfts.tsx:34]: got it

File: src/Nfts/Nfts.tsx (hunk 12 above)
Comment (banklesss) [reply src/Nfts/Nfts.tsx:56]: I created a ticket to implement the rescoping https://emurgo.atlassian.net/browse/YOMO-518

Thanks for the suggestion.

**Comment by banklesss** [comment]
Rebase here https://github.com/Emurgo/yoroi-mobile/pull/2424

**Pull request closed without merging**